stackslib/src/chainstate/burn/db/sortdb.rs

```diff
@@ -1859,7 +1854,13 @@ impl<'a> SortitionHandleConn<'a> {
                 e
             })?;
 
-        if ch_sn.block_height + u64::from(self.context.pox_constants.reward_cycle_length)
+        if ch_sn.block_height
+            + u64::from(
+                self.context
+                    .pox_constants
+                    .reward_cycle_length
+                    .saturating_mul(2),
+            )
```

Comment:
This has to do with the unsoundness of how we deal with reward cycles -- sometimes, we think of them as starting at reward cycle index 1, and other times, at reward cycle index 0.  Aggregate public keys take effect at index 1, but a tenure at index 0 belongs to the same reward cycle of the tenure at index 1, and will be verifiable by two different aggregate public keys.  The tenure at index 0 would be signed by the private shares for the aggregate public key chosen not in the last cycle, but the cycle prior to that.  Index 1 and up would be signed by the private shares generated in the prior cycle.  This is why you need to accept blocks that are verifiable by an aggregate public key from the last two cycles.

For example, suppose we have reward cycles of length 10 and a prepare phase of 5 blocks.  The tenure at sortition height 30 would be signed with the aggregate public key chosen at block 16, which is over 10 blocks in the past.  The tenure at sortition height 31 would be signed by the aggregate public key chosen at block 26.  Since we are able to obtain the tenure at sortition height 30 before processing the PoX anchor block mined in sortition 26, I altered this code to look back two reward cycles.

Reply:
>  Aggregate public keys take effect at index 1, but a tenure at index 0 belongs to the same reward cycle of the tenure at index 1, and will be verifiable by two different aggregate public keys.

This doesn't sound ideal, and it doesn't sound like something that needs to be taken as a given. Currently, the way that aggregate public key checks work is as follows:

1. The relayer's block acceptance is the only place the aggregate public key is checked (this needs to change, obviously), but it does:

```rust
        let Ok(aggregate_public_key) =
            NakamotoChainState::get_aggregate_public_key(chainstate, &sortdb, sort_handle, &block)
        else {
            warn!("Failed to get aggregate public key. Will not store or relay";
                "stacks_block_hash" => %block.header.block_hash(),
                "consensus_hash" => %block.header.consensus_hash,
                "burn_height" => block.header.chain_length,
                "sortition_height" => block_sn.block_height,
            );
            return Ok(false);
        };
```

Before invoking the block acceptance function, passing in the aggregate public key.

2. `NakamotoChainstate::get_aggregate_public_key()` itself looks at the block's consensus hash to figure out its burn height. Based on that, it invokes `block_height_to_reward_cycle`, which returns `(effective_height / reward_cycle_len)` to figure out the reward cycle ID.

So that sounds like only one aggregate key is ever checked for a burn block height (which sounds correct), and it uses the `effective_height / reward_cycle_len` calculation.

So maybe the `mul(2)` isn't necessary? But either way, it deserves more comments in the code, and an update to the rustdoc to indicate the change.

Reply:
I did some testing, and the `mul(2)` is an overestimate, but I can confirm that  removing the `mul(2)` breaks the test due to a lack of an aggregate public key.  The search window size is really `reward_cycle_len + prepare_len`, because the last block in reward cycle N+1 would be signed by the private shares for the aggregate public key chosen in the first prepare-phase block in cycle N.

I misspoke above.  I didn't mean to say that a block could be signed with one key _OR_ another.  It's signed with the key from its reward cycle's prepare phase, which is can be mined more than `reward_cycle_blocks` sortitions in the past.

Reply:
I opened an issue to track verifying the aggregate signature on `append_block`: https://github.com/stacks-network/stacks-core/issues/4571